### DISCUSSION TITLE
Bug - Unittest indefinitely hangs when using BaseHTTPMiddleware, sync router and quit(). Needs MainThread _daemonic hack.

### DISCUSSION THREAD

**Discussion by arrman-saha** [encode/starlette#2537] [Potential Issue]
When a BaseHTTPMiddleware is used, and we want to quit the application in the router everything works as expected when deploying the app. However unit testcases permanently hang if we use a sync router. To make the testcases not hang we need a hack - `setattr(threading.main_thread(), "_daemonic", True)`. But this hack is not needed for async router.

I've tried looking for similar issues but can't find something similar.

I originally thought this was a [fastapi issue](https://github.com/tiangolo/fastapi/discussions/11254), but seems like this is a starlette issue. 

The following code will hang indefinitely after testcases have passed, unless either the `async def` part is used or  `setattr` part is used.

```python
import unittest
from starlette.applications import Starlette
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.middleware import Middleware
from starlette.testclient import TestClient
from starlette.routing import Route

# async def just_quit(_):  # Async def does not need the MainThread _daemonic hack
def just_quit(_):
    quit(1)

class NothingMiddleware(BaseHTTPMiddleware):
    def __init__(self,app,):
        super().__init__(app)
    async def dispatch(self, request, call_next):
        return await call_next(request)

app = Starlette(debug=False, routes=[Route("/quit", just_quit)], middleware=[Middleware(NothingMiddleware)])

class TestQuitEndpoint(unittest.TestCase):

    def test_SystemExit(self):
        # Non async def needs this MainThread _daemonic hack.
        # import threading
        # setattr(threading.main_thread(), "_daemonic", True)
        # Testcase
        with self.assertRaises(RuntimeError):
            with TestClient(app) as client:
                with self.assertRaises(SystemExit):
                    client.get("/quit")

if __name__ == '__main__':
    unittest.main()
```

**Comment by arrmansa** [comment]
Hi, looks like anyio worker threads were getting stuck because BaseException was not handled in BaseHTTPMiddleware. Can I open a PR for this ? 
https://github.com/encode/starlette/compare/master...arrmansa:starlette:master 

(The ask account is different because I accidentally started discussion from my work account. I'm not trying to spam the discussions with an alt or anything.)
